File: src/main/java/com/jitong/im/media/ImageNormalizer.java
package com.jitong.im.media;

import com.jitong.im.platform.error.ApiErrorDefinition;

import javax.imageio.ImageIO;
import javax.imageio.ImageReader;
import javax.imageio.stream.ImageInputStream;
import java.awt.AlphaComposite;
import java.awt.Color;
import java.awt.Graphics2D;
import java.awt.RenderingHints;
import java.awt.image.BufferedImage;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.security.MessageDigest;
import java.util.Iterator;
import java.util.Locale;

final class ImageNormalizer {

    static final int MAX_INPUT_BYTES = 10 * 1024 * 1024;
    static final int MAX_OUTPUT_BYTES = 5 * 1024 * 1024;
    static final int MAX_LONG_EDGE = 2048;
    static final int THUMBNAIL_LONG_EDGE = 320;
    static final long MAX_DECODE_PIXELS = 20_000_000L;

    private ImageNormalizer() {
    }

    static NormalizedImage normalize(byte[] input) {
        if (input == null || input.length == 0) {
            throw new MediaException(ApiErrorDefinition.MEDIA_INVALID);
        }
        if (input.length > MAX_INPUT_BYTES) {
            throw new MediaException(ApiErrorDefinition.MEDIA_TOO_LARGE);
        }

        BufferedImage decoded = decode(input);
        BufferedImage normalized = scale(decoded, MAX_LONG_EDGE);
        byte[] original = encodeJpeg(normalized);
        if (original.length > MAX_OUTPUT_BYTES) {
            throw new MediaException(ApiErrorDefinition.MEDIA_TOO_LARGE);
        }
        byte[] thumbnail = encodeJpeg(scale(decoded, THUMBNAIL_LONG_EDGE));
        return new NormalizedImage(
                original,
                thumbnail,
                normalized.getWidth(),
                normalized.getHeight(),
                "image/jpeg",
                sha256(original));
    }

    private static BufferedImage decode(byte[] input) {
        try (ImageInputStream imageInput = ImageIO.createImageInputStream(
                new ByteArrayInputStream(input))) {
            if (imageInput == null) {
                throw new MediaException(ApiErrorDefinition.MEDIA_INVALID);
            }
            Iterator<ImageReader> readers = ImageIO.getImageReaders(imageInput);
            if (!readers.hasNext()) {
                throw new MediaException(ApiErrorDefinition.MEDIA_INVALID);
            }
            ImageReader reader = readers.next();
            try {
                reader.setInput(imageInput, true, true);
                String format = reader.getFormatName().toLowerCase(Locale.ROOT);
                if (!format.equals("jpeg") && !format.equals("jpg")
                        && !format.equals("png") && !format.equals("gif")) {
                    throw new MediaException(ApiErrorDefinition.MEDIA_INVALID);
                }
                int width = reader.getWidth(0);
                int height = reader.getHeight(0);
                if (width <= 0 || height <= 0
                        || ((long) width * height) > MAX_DECODE_PIXELS) {
                    throw new MediaException(ApiErrorDefinition.MEDIA_DIMENSIONS_TOO_LARGE);
                }
                BufferedImage decoded = reader.read(0);
                if (decoded == null) {
                    throw new MediaException(ApiErrorDefinition.MEDIA_INVALID);
                }
                return decoded;
            } finally {
                reader.dispose();
            }
        } catch (MediaException exception) {
            throw exception;
        } catch (IOException | RuntimeException exception) {
            throw new MediaException(ApiErrorDefinition.MEDIA_INVALID, exception);
        }
    }

    private static BufferedImage scale(BufferedImage source, int maxLongEdge) {
        double scale = Math.min(1d, maxLongEdge / (double) Math.max(
                source.getWidth(),
                source.getHeight()));
        int width = Math.max(1, (int) Math.round(source.getWidth() * scale));
        int height = Math.max(1, (int) Math.round(source.getHeight() * scale));
        BufferedImage target = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
        Graphics2D graphics = target.createGraphics();
        graphics.setComposite(AlphaComposite.Src);
        graphics.setColor(Color.WHITE);
        graphics.fillRect(0, 0, width, height);
        graphics.setRenderingHint(
                RenderingHints.KEY_INTERPOLATION,
                RenderingHints.VALUE_INTERPOLATION_BILINEAR);
        graphics.setRenderingHint(
                RenderingHints.KEY_RENDERING,
                RenderingHints.VALUE_RENDER_QUALITY);
        graphics.drawImage(source, 0, 0, width, height, null);
        graphics.dispose();
        return target;
    }

    private static byte[] encodeJpeg(BufferedImage image) {
        try {
            ByteArrayOutputStream output = new ByteArrayOutputStream();
            if (!ImageIO.write(image, "jpg", output)) {
                throw new MediaException(ApiErrorDefinition.MEDIA_INVALID);
            }
            return output.toByteArray();
        } catch (IOException exception) {
            throw new MediaException(ApiErrorDefinition.MEDIA_INVALID, exception);
        }
    }

    private static String sha256(byte[] bytes) {
        try {
            return java.util.HexFormat.of().formatHex(
                    MessageDigest.getInstance("SHA-256").digest(bytes));
        } catch (Exception exception) {
            throw new IllegalStateException(exception);
        }
    }

    record NormalizedImage(
            byte[] original,
            byte[] thumbnail,
            int width,
            int height,
            String contentType,
            String sha256
    ) {
    }
}

File: desktop-app/src/main/kotlin/com/jitong/im/desktop/media/ImageNormalizer.kt
package com.jitong.im.desktop.media

import java.awt.AlphaComposite
import java.awt.Color
import java.awt.Graphics2D
import java.awt.RenderingHints
import java.awt.image.BufferedImage
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.security.MessageDigest
import java.util.HexFormat
import javax.imageio.ImageIO
import javax.imageio.ImageWriteParam

internal object ImageNormalizer {
    const val MAX_INPUT_BYTES = 10 * 1024 * 1024
    const val MAX_OUTPUT_BYTES = 5 * 1024 * 1024
    const val MAX_LONG_EDGE = 2048
    private const val MAX_DECODE_PIXELS = 20_000_000L

    fun normalize(source: ByteArray): ByteArray {
        require(source.isNotEmpty() && source.size <= MAX_INPUT_BYTES) {
            "Image input is too large"
        }
        val decoded = ImageIO.read(ByteArrayInputStream(source))
            ?: throw IllegalArgumentException("Image cannot be decoded")
        require(
            decoded.width > 0
                && decoded.height > 0
                && decoded.width.toLong() * decoded.height <= MAX_DECODE_PIXELS) {
            "Image dimensions are too large"
        }
        val normalized = scale(decoded)
        return try {
            encodeJpeg(normalized)
        } finally {
            normalized.flush()
            decoded.flush()
        }
    }

    private fun scale(source: BufferedImage): BufferedImage {
        val scale = minOf(
            1.0,
            MAX_LONG_EDGE.toDouble() / maxOf(source.width, source.height))
        val width = maxOf(1, (source.width * scale).toInt())
        val height = maxOf(1, (source.height * scale).toInt())
        return BufferedImage(width, height, BufferedImage.TYPE_INT_RGB).also { target ->
            val graphics = target.createGraphics()
            graphics.composite = AlphaComposite.Src
            graphics.color = Color.WHITE
            graphics.fillRect(0, 0, width, height)
            graphics.setRenderingHint(
                RenderingHints.KEY_INTERPOLATION,
                RenderingHints.VALUE_INTERPOLATION_BILINEAR)
            graphics.setRenderingHint(
                RenderingHints.KEY_RENDERING,
                RenderingHints.VALUE_RENDER_QUALITY)
            graphics.drawImage(source, 0, 0, width, height, null)
            graphics.dispose()
        }
    }

    private fun encodeJpeg(image: BufferedImage): ByteArray {
        var quality = 0.9f
        while (quality >= 0.4f) {
            val output = ByteArrayOutputStream()
            val writer = ImageIO.getImageWritersByFormatName("jpg").asSequence().firstOrNull()
                ?: throw IllegalStateException("JPEG writer is unavailable")
            try {
                writer.output = ImageIO.createImageOutputStream(output)
                val params = writer.defaultWriteParam.apply {
                    compressionMode = ImageWriteParam.MODE_EXPLICIT
                    compressionQuality = quality
                }
                writer.write(null, javax.imageio.IIOImage(image, null, null), params)
            } catch (exception: IOException) {
                throw IllegalArgumentException("Image cannot be encoded", exception)
            } finally {
                writer.dispose()
            }
            val bytes = output.toByteArray()
            if (bytes.size <= MAX_OUTPUT_BYTES) return bytes
            quality -= 0.1f
        }
        throw IllegalArgumentException("Image is too large after normalization")
    }

    fun sha256(source: ByteArray): String =
        HexFormat.of().formatHex(MessageDigest.getInstance("SHA-256").digest(source))
}
